### PR TITLE
{devel} [goolf-1.5.14-Python-2.7.6] PyQt4 SIP (WIP)

### PR DESCRIPTION
**Pull request by Bart-Ver**
This PR needs easyblock PR ~~https://github.com/hpcugent/easybuild-easyblocks/pull/540~~

SIP works and installs well.

PyQt doesn't make it through the sanity check, as it wants to import the module pyqt, but the module name is PyQt4. Using the 'options' doesn't seem to work.
Debug output: https://gist.github.com/Bart-CER/4b59e43db11f33e46b8d.

### CODE REVIEW

File: easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+sources = ['%(name)s-x11-gpl-%(version)s.tar.gz']
+source_urls = ['http://sourceforge.net/projects/pyqt/files/PyQt4/PyQt-4.11.3']
+
+python = 'Python'
+pythonver = '2.7.6'
+pythonshortver = '.'.join(pythonver.split('.')[0:2])
+versionsuffix = '-%s-%s' % (python, pythonver)
+
+dependencies = [(python, pythonver),
```
Comment (wpoely86) [inline easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb]: Format this as:
```python
dependencies = [
     (python, pythonver),
     ('SIP','4.16.4',versionsuffix),
     ('Qt','4.8.4')
]
```

**Comment by wpoely86** [comment]
ok to test

**Comment by wpoely86** [comment]
This shouldn't happen:
```
== 2015-01-29 10:32:49,740 main.EasyConfig DEBUG Ignoring unknown config option modulename (value: PyQt4)
```

**Comment by Bart-CER** [comment]
This will not work without framework PR  #1153.
It does work when options in the config file are possible.

It seems I cannot use intalldir and pythonvershort at the same time in

configopts = " configure-ng.py --confirm-license --destdir=%(installdir)s/lib/python2.7/site-packages --no-sip-files"

**Comment by boegel** [comment]
@wpoely86: it's a framework bug, see https://github.com/hpcugent/easybuild-framework/pull/1153

But I guess we need a better way than having to use `options = ...`, since `options` is way too generic.

**Comment by boegel** [comment]
@Bart-CER: to use both `%(installdir)s` and `pythonshortver` at the same time, you need to escape the `%` in `%(installdir)s`:

```python
configopts = " configure-ng.py --confirm-license --destdir=%%(installdir)s/lib/python%s/site-packages --no-sip-files" % pythonshortver
```

File: easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+pythonshortver = '.'.join(pythonver.split('.')[0:2])
+versionsuffix = '-%s-%s' % (python, pythonver)
+
+dependencies = [
+        (python, pythonver),
+        ('SIP','4.16.4',versionsuffix),
+        ('Qt','4.8.4')
+]
+
+configopts = " configure-ng.py --confirm-license --destdir=%(installdir)s/lib/python2.7/site-packages --no-sip-files"
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb]:
```python
configopts = " configure-ng.py --confirm-license --destdir=%%(installdir)s/lib/python%s/site-packages --no-sip-files" % pythonshortver
```

File: easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+python = 'Python'
+pythonver = '2.7.6'
+pythonshortver = '.'.join(pythonver.split('.')[0:2])
+versionsuffix = '-%s-%s' % (python, pythonver)
+
+dependencies = [(python, pythonver)]
+
+preconfigopts = 'mkdir -p %(installdir)s/bin %(installdir)s/lib/python2.7/site-packages/SIP %(installdir)s/include %(installdir)s/sip && '
+configopts = " configure.py -b %(installdir)s/bin -d %(installdir)s/lib/python2.7/site-packages/SIP -e %(installdir)s/include -v %(installdir)s/sip"
```
Comment (boegel) [inline easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb]: avoid too long lines, use `configopts +=` to break it up

don't hardcode `2.7`, use `pythonshortver` and escape `%` where you need to (`%(installdir)s`)

File: easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+PyQt is a set of Python v2 and v3 bindings for Digia's Qt application framework 
+and runs on all platforms supported by Qt including Windows, MacOS/X and Linux. 
+PyQt5 supports Qt v5. PyQt4 supports Qt v4 and will build against Qt v5. 
+The bindings are implemented as a set of Python modules and contain over 620 classes. 
+"""
+
+toolchain = {'name': 'goolf', 'version': '1.5.14'}
+
+sources = ['%(name)s-x11-gpl-%(version)s.tar.gz']
+source_urls = ['http://sourceforge.net/projects/pyqt/files/PyQt4/PyQt-4.11.3']
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb]: use `%(version)s`

File: easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+python = 'Python'
+pythonver = '2.7.6'
+pythonshortver = '.'.join(pythonver.split('.')[0:2])
+versionsuffix = '-%s-%s' % (python, pythonver)
+
+dependencies = [
+        (python, pythonver),
+        ('SIP','4.16.4',versionsuffix),
+        ('Qt','4.8.4')
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb]: indent with (just) 4 spaces (no tabs)

File: easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+]
+
+configopts = " configure-ng.py --confirm-license --destdir=%(installdir)s/lib/python2.7/site-packages --no-sip-files" 
+
+options = {'modulename': 'PyQt4'}
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb]: this looks too generic, we should probably change that to something more specific (via https://github.com/hpcugent/easybuild-easyconfigs/pull/1153)

File: easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+source_urls = ['http://sourceforge.net/projects/pyqt/files/sip/sip-4.16.4']
+
+python = 'Python'
+pythonver = '2.7.6'
+pythonshortver = '.'.join(pythonver.split('.')[0:2])
+versionsuffix = '-%s-%s' % (python, pythonver)
+
+dependencies = [(python, pythonver)]
+
+preconfigopts = 'mkdir -p %(installdir)s/bin %(installdir)s/lib/python2.7/site-packages/SIP %(installdir)s/include %(installdir)s/sip && '
```
Comment (boegel) [inline easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb]: same here, don't hardcode `2.7`

File: easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+configopts = " configure.py -b %(installdir)s/bin -d %(installdir)s/lib/python2.7/site-packages/SIP -e %(installdir)s/include -v %(installdir)s/sip" 
+buildininstalldir=True
+
+modextrapaths = {
+    'CPATH':'include',
+    'LD_LIBRARY_PATH':'lib/python%s/site-packages/SIP'%pythonshortver,
+    'LIBRARY_PATH':'lib/python%s/site-packages/SIP'%pythonshortver,
+    'PYTHONPATH':'lib/python%s/site-packages/SIP'%pythonshortver,
```
Comment (boegel) [inline easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb]: code style nitpicking: spaces around `%` and after `:`

File: easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb (hunk 6 above)
Comment (boegel) [inline easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb]: actually, we should take the approach discussed in https://github.com/hpcugent/easybuild-framework/issues/991...

but that's more long-term, this should work for now

File: easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb (hunk 5 above)
Comment (boegel) [inline easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb]: also, bump version to Qt 4.8.6 (which is already available in develop, 4.8.4 is not for this toolchain)

File: easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+for the Qt toolkit, but can be used to create bindings for any C or C++ library. 
+"""
+
+toolchain = {'name': 'goolf', 'version': '1.5.14'} 
+
+sources = [SOURCELOWER_TAR_GZ]
+source_urls = ['http://sourceforge.net/projects/pyqt/files/sip/sip-4.16.4']
+
+python = 'Python'
+pythonver = '2.7.6'
```
Comment (boegel) [inline easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb]: bump to 2.7.9, easyconfig is already there in develop with this toolchain?

**Comment by Bart-CER** [comment]
I added 2 dependencies, as this is what we have installed here. 

This doesn’t work for me, still an issue with those options:

File "/home/bver018/easy/lib/python2.6/site-packages/easybuild_easyblocks-2.0.0dev-py2.6.egg/easybuild/easyblocks/generic/pythonpackage.py", line 234, in sanity_check_step
    return super(PythonPackage, self).sanity_check_step(*args, **kwargs)
TypeError: sanity_check_step() got multiple values for keyword argument 'exts_filter'

File: easybuild/easyconfigs/p/Python/Python-2.7.6-goolf-1.5.14.eb [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    }),
+    ('Bottleneck', '0.8.0', {
+        'source_urls': ['https://pypi.python.org/packages/source/B/Bottleneck/'],
+    }),
+    ('pandas', '0.13.1', {
+        'source_urls': ['https://pypi.python.org/packages/source/p/pandas/'],
+    }),
+]
+
+osdependencies = [('openssl-devel', 'libssl-dev')]
```
Comment (boegel) [inline easybuild/easyconfigs/p/Python/Python-2.7.6-goolf-1.5.14.eb]: move this up, right below `dependencies` (empty line in between)

File: easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+dependencies = [
+    (python, pythonver),
+    ('SIP','4.16.4',versionsuffix),
+    ('Qt','4.8.4')
+]
+
+
+configopts = " configure-ng.py --confirm-license "
+    += "--destdir=%%(installdir)s/lib/python%s/site-packages " %pythonshortver
+    += "--no-sip-files"
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb]: you need to use `configopts +=` to get this to work, right?

also: space after `%` in `%pythonshortver`

File: easybuild/easyconfigs/p/Python/Python-2.7.6-goolf-1.5.14.eb [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+dependencies = [
+    ('bzip2', '1.0.6'),        # for bz2 package
+    ('zlib', '1.2.8'),
+    ('libreadline', '6.2'),
+    ('ncurses', '5.9'),
+    ('LLVM', '3.2'),           # for Numba
+#   ('OpenSSL', '1.0.1f'),     # OS dependency should be preferred for security reasons
+    ('libjpeg-turbo','1.3.0'), # for PIL/Pillow
+    ('libpng','1.6.10'),       # for PIL/Pillow
+    ('freetype', '2.5.3'),     # for PIL/Pillow
```
Comment (boegel) [inline easybuild/easyconfigs/p/Python/Python-2.7.6-goolf-1.5.14.eb]: we try to stay away from adding dependencies for specific Python packages; ideally, these should be installed via a separate easyconfig

File: easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+description = """
+SIP is a tool that makes it very easy to create Python bindings for C and C++ 
+libraries. It was originally developed to create PyQt, the Python bindings 
+for the Qt toolkit, but can be used to create bindings for any C or C++ library. 
+"""
+
+toolchain = {'name': 'goolf', 'version': '1.5.14'} 
+
+sources = [SOURCELOWER_TAR_GZ]
+source_urls = ['http://sourceforge.net/projects/pyqt/files/sip/sip-4.16.4']
```
Comment (boegel) [inline easybuild/easyconfigs/s/SIP/SIP-4.16.4-goolf-1.5.14-Python-2.7.6.eb]: don't hardcode the version here, use `%(version)s`

File: easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb (hunk 11 above)
Comment (Bart-CER) [inline easybuild/easyconfigs/p/PyQt/PyQt-4.11.3-goolf-1.5.14-Python-2.7.6.eb]: no, configopts = works.

File: easybuild/easyconfigs/p/Python/Python-2.7.6-goolf-1.5.14.eb (hunk 12 above)
Comment (Bart-CER) [inline easybuild/easyconfigs/p/Python/Python-2.7.6-goolf-1.5.14.eb]: This is the config file we used, I can delete all the dependencies if you think that's better?

**Comment by Bart-CER** [comment]
There is still an issue with the sanity check using the ConfigureMakePythonPackage block:

ERROR: EasyBuild crashed with an error (at easybuild/main.py:135 in build_and_install_software): Traceback (most recent call last):
  File "/home/bver018/easy/lib/python2.6/site-packages/easybuild_framework-2.0.0dev-py2.6.egg/easybuild/main.py", line 109, in build_and_install_software
    (ec_res['success'], app_log, err) = build_and_install_one(ec, orig_environ)
  File "/home/bver018/easy/lib/python2.6/site-packages/easybuild_framework-2.0.0dev-py2.6.egg/easybuild/framework/easyblock.py", line 1877, in build_and_install_one
    result = app.run_all_steps(run_test_cases=run_test_cases)
  File "/home/bver018/easy/lib/python2.6/site-packages/easybuild_framework-2.0.0dev-py2.6.egg/easybuild/framework/easyblock.py", line 1818, in run_all_steps
    self.run_step(stop_name, step_methods, skippable=skippable)
  File "/home/bver018/easy/lib/python2.6/site-packages/easybuild_framework-2.0.0dev-py2.6.egg/easybuild/framework/easyblock.py", line 1704, in run_step
    m(self)
  File "/home/bver018/easy/lib/python2.6/site-packages/easybuild_framework-2.0.0dev-py2.6.egg/easybuild/framework/easyblock.py", line 1788, in <lambda>
    ('sanitycheck', 'sanity checking', [lambda x: x.sanity_check_step()], False),
  File "/home/bver018/easy/lib/python2.6/site-packages/easybuild_easyblocks-2.0.0dev-py2.6.egg/easybuild/easyblocks/generic/configuremakepythonpackage.py", line 77, in sanity_check_step
    return PythonPackage.sanity_check_step(self, EXTS_FILTER_PYTHON_PACKAGES, *args, **kwargs)
  File "/home/bver018/easy/lib/python2.6/site-packages/easybuild_easyblocks-2.0.0dev-py2.6.egg/easybuild/easyblocks/generic/pythonpackage.py", line 234, in sanity_check_step
    return super(PythonPackage, self).sanity_check_step(*args, **kwargs)
TypeError: sanity_check_step() got multiple values for keyword argument 'exts_filter'

**Comment by boegel** [comment]
@Bart-CER: you're not using the latest updated easyblock from https://github.com/hpcugent/easybuild-easyblocks/pull/540, please recheck

**Comment by Bart-CER** [comment]
@boegel  Sorry, didn't realise you changed 2 files in the generic blocks. It's working perfectly now.

**Comment by pescobar** [comment]
I also need PyQT. This easyconfig in combination with this easyblock https://github.com/hpcugent/easybuild-easyblocks/pull/540 seems to be working fine for me. At least the installation has completed correctly, I am going to do some more testing

Thank you very much @Bart-CER

@boegel it would be really nice to get this one merged in

**Comment by boegel** [comment]
Jenkins: test this please

**Comment by boegel** [comment]
A PyQt and SIP easyconfig using the latest Python version is merged in via #1322, we should look into syncing this PR accordingly.

The Qt version used in the PyQt easyconfig is different, and the Python 2.7.6 easyconfig also contains bumped versions of extensions, and new extensions...

**Comment by migueldiascosta** [comment]
@Bart-Ver, this PR is being closed for the following reason(s): no activity for > 6 months.
Please don't hesitate to reopen this PR or add a comment if you feel this contribution is still relevant.
For more information on our policy w.r.t. closing PRs, see https://easybuild.readthedocs.io/en/latest/Contributing.html#why-a-pull-request-may-be-closed-by-a-maintainer

**Pull request closed without merging**